### PR TITLE
[POAE7-335] libpmemblk backend and libpmemkv metastore implementation

### PR DESCRIPTION
**Pull request by yeyuqiang**

### CODE REVIEW

File: oap-common/src/native/libpmemblk/com_intel_oap_common_unsafe_PMemBlockPlatform.cpp [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+ * Method:    read
+ * Signature: (Ljava/nio/ByteBuffer;I)V
+ */
+JNIEXPORT void JNICALL Java_com_intel_oap_common_unsafe_PMemBlockPlatform_read
+  (JNIEnv *env, jclass clazz, jbyteArray jbuf, jint jindex) {
+
+  jbyte* buf = env->GetByteArrayElements(jbuf, 0);
+  int index = (int) jindex;
+
+  if(pmemblk_read(pbp, buf, index) < 0) {
```
Comment (jikunshang) [inline oap-common/src/native/libpmemblk/com_intel_oap_common_unsafe_PMemBlockPlatform.cpp:87]: This API needs an extra copy like vmemcache, we can't use a libpmemblk address, right? If yes, there may be some overhead.

File: oap-common/src/test/java/com/intel/oap/common/unsafe/PMemBlockPlatformTest.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        }
+    }
+
+    @AfterClass
+    public static void close() {
+        PMemBlockPlatform.close();
+    }
+
+    private byte[] writeBlock(int index) {
+        byte[] bytesToWrite = new byte[ELEMENT_SIZE];
```
Comment (jikunshang) [inline oap-common/src/test/java/com/intel/oap/common/unsafe/PMemBlockPlatformTest.java:43]: Do we support multi thread read/write?

File: oap-common/src/main/java/com/intel/oap/common/storage/pmemblk/PMemBlkChunkReader.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        this.chunkSize = pMemManager.getChunkSize();
+    }
+
+    @Override
+    protected int readFromPMem(PMemPhysicalAddress pMemPhysicalAddress, ByteBuffer data) {
+        PMemBlkPhysicalAddress pMemBlkPhysicalAddress = (PMemBlkPhysicalAddress) pMemPhysicalAddress;
+        int index = pMemBlkPhysicalAddress.getIndex();
+        byte[] buffer = new byte[chunkSize];
+        PMemBlockPlatform.read(buffer, index);
+        data.put(buffer);
```
Comment (yma11) [inline oap-common/src/main/java/com/intel/oap/common/storage/pmemblk/PMemBlkChunkReader.java:25]: Why should here always read whole data of this index? What will happen if chunkSize>data.size()?

File: oap-common/src/main/java/com/intel/oap/common/storage/pmemblk/PMemBlkPhysicalAddress.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package com.intel.oap.common.storage.pmemblk;
+
+import com.intel.oap.common.storage.stream.PMemPhysicalAddress;
+
+public class PMemBlkPhysicalAddress implements PMemPhysicalAddress {
+
+    private int index;
```
Comment (yma11) [inline oap-common/src/main/java/com/intel/oap/common/storage/pmemblk/PMemBlkPhysicalAddress.java:7]: It makes sense we only track index of this pmemblk when it's not the last trunk of the stream. But for the last chunk, don't we need to track the actual size used of this pmemblk?

File: oap-common/src/test/java/com/intel/oap/common/unsafe/PMemBlockPlatformTest.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    @Test
+    public void testWriteMultipleBlock() {
+        for (int i = 0; i < 32; i++) {
+            byte[] writtenBlock = writeBlock(i);
+            byte[] readBlock = readBlock(i);
+            assertArrayEquals(writtenBlock, readBlock);
+        }
+    }
```
Comment (yma11) [inline oap-common/src/test/java/com/intel/oap/common/unsafe/PMemBlockPlatformTest.java:69]: I think you only test the boundary cases that data size is X*chunkSize,  we need additional one to cover X*chunkSize + Y case, right?

File: oap-common/src/native/libpmemblk/com_intel_oap_common_unsafe_PMemBlockPlatform.cpp (hunk 1 above)
Comment (yeyuqiang) [reply oap-common/src/native/libpmemblk/com_intel_oap_common_unsafe_PMemBlockPlatform.cpp:87]: Should be no extra copy. pmemblk_read will handle the copy from address(determine by index) to arrays of block (buf)

File: oap-common/src/test/java/com/intel/oap/common/unsafe/PMemBlockPlatformTest.java (hunk 2 above)
Comment (yeyuqiang) [reply oap-common/src/test/java/com/intel/oap/common/unsafe/PMemBlockPlatformTest.java:43]: Yes

File: oap-common/src/main/java/com/intel/oap/common/storage/pmemblk/PMemBlkChunkReader.java (hunk 3 above)
Comment (yeyuqiang) [reply oap-common/src/main/java/com/intel/oap/common/storage/pmemblk/PMemBlkChunkReader.java:25]: pmemblk will always put a buf with chunkSize otherwise it will fail.

File: oap-common/src/test/java/com/intel/oap/common/unsafe/PMemBlockPlatformTest.java (hunk 5 above)
Comment (yeyuqiang) [reply oap-common/src/test/java/com/intel/oap/common/unsafe/PMemBlockPlatformTest.java:69]: For libpmemblk, the case should be (x+y)chunksize. Then testWriteBlockExceedMaximum is the expected test case

File: oap-common/src/main/java/com/intel/oap/common/storage/pmemblk/PMemBlkPhysicalAddress.java (hunk 4 above)
Comment (yeyuqiang) [reply oap-common/src/main/java/com/intel/oap/common/storage/pmemblk/PMemBlkPhysicalAddress.java:7]: I think metadata mechanism has already handle this problem.

Comment (yma11) [reply oap-common/src/main/java/com/intel/oap/common/storage/pmemblk/PMemBlkPhysicalAddress.java:7]: Metafooter only record total trunks of this stream but not total size. There is no method to get actual size of last trunk.

Comment (yeyuqiang) [reply oap-common/src/main/java/com/intel/oap/common/storage/pmemblk/PMemBlkPhysicalAddress.java:7]: memory stats is tracking on total size and usage. I should return the buffer size in readFromPMem for the last trun is < chunksize

Comment (yeyuqiang) [reply oap-common/src/main/java/com/intel/oap/common/storage/pmemblk/PMemBlkPhysicalAddress.java:7]: and record the real chunk length in PhysicalAddress. Does it make sense?

Comment (yma11) [reply oap-common/src/main/java/com/intel/oap/common/storage/pmemblk/PMemBlkPhysicalAddress.java:7]: I think so. I did that in memkind. and also you may need to rethink about the write/read part, that you may read a whole trunk but can't just put all of them into returned buffer?

**Comment by yeyuqiang** [comment]
any other further comments? If not, I will merge it and open new PR for integration
